rfcs/012-fdb-reduce.md

```diff
@@ -0,0 +1,190 @@
+
+See figure 2 for an example of the listed keys added to a skip list.
+
+![figure 2: The emitted reduce keys added to a reduce skip list](images/SkExample2.png)
+*figure 2:* The emitted reduce keys added to a reduce skip lis
+
+The optimal number of levels and the correct distribution algorithm is unknown at this stage, these settings will be configurable so that we can test and determine the best fit.
+
+#### Query
+
+From the above figure 2, we can see that for a reduce query with `group = true` then level 0 will be used to return all the exact keys.
+And for a query with `group_level = 0` then the highest level can be used. 
+If a `group_level > 1` is set for a reduce query, we need to traverse the skip list and aggregate the results before returning the results to the user.
+
+For example, using figure 2, with a reduce query of `group_level = 2`. We start at Level 4, we would traverse down to level 3, compare Node 0 and Node [2018, 03, 1]. They are not the same key for `group_level = 2` so we need to move to a lower level. We would perform the same action of comparing the current node with the next node to see if they are the same key until we find matching nodes or we get to level 0. In this case, we would reach level 0 and return Node [2017, 03, 1]. At Node [2017, 04, 1] it is possible to go back up to level 2 where we can compare [2017, 04, 1] and [2017, 04, 15]. Given they are the same, we can continue on level 1 and collect all the [2017, 04, x] keys. Once we have collected all of those. Those keys are run through a re-reduce before being returned. This process of comparing the current node with the next node would continue onwards. Each time comparing the current node with the next node. If they are the same, compare the Nodes at the next Level above until we reach a level where the keys are not equal. At that point, we would use the node values at the level just below that to move across, collecting all the equal keys to be run through a reducer before being returned.
+
+![figure 3: Traversing flow to return results for `group_level = 2`](images/SkExample3.png)
+*Figure 3:* Traversing flow to return results for `group_level = 2`
+
+A query with startkey/endkey would follow a similar process.  Start at the highest level traversing across until we exceed the startkey, then move down until we find the startkey or nearest node to it. Then follow the above process of traversing the skip list to return all results until we reach a node that is greater or equal to the endkey.
+
+#### Update
+
+To update the reduce index, we will use the same map id index that keeps track of what keys are associated with a document. When a document is deleted. The previous keys will be removed from level 0. If the reduce function is `_sum` or `_count`, an atomic update is then performed for all Node's above Level 0 that would have included the values for the keys deleted.  
```

Comment:
"When a document is deleted." exists adrift from context, which sentence is it part of?